### PR TITLE
test(loss_functions): add XGBoost 3.2 2D input regression tests

### LINKED ISSUE #865: feat(compat): add XGBoost 3.2 compatibility for custom objectives
## Background

XGBoost 3.2 (Feb 2026) introduced significant changes to multi-target tree support with vector-leaf and changes to how multi-output data flows through the sklearn interface's `_objective_decorator`. This broke OpenSTEF's custom objective functions.

As a **temporary fix**, XGBoost was pinned to `<3.2` in #864. This issue tracks the proper fix.

## Root Cause

The custom objectives (`arctan_loss_multi_objective`, `pinball_loss_multi_objective`) in `packages/openstef-models/src/openstef_models/utils/loss_functions.py` assume XGBoost passes a **1D flattened array** to the objective:

```python
n_items = len(y_true)
n_quantiles = len(quantiles)
n_rows = n_items // n_quantiles          # ← fragile assumption
y_pred = np.reshape(y_pred, (n_rows, n_quantiles))
y_true = np.reshape(y_true, (n_rows, n_quantiles))
```

Similarly in `mean_pinball_loss` (evaluation function):
```python
y_pred = np.reshape(y_pred, [-1, n_quantiles])
```

XGBoost 3.2 may pass **2D `(n_samples, n_quantiles)` arrays** directly to the custom objective when using `multi_strategy="one_output_per_tree"`, making the manual reshape unnecessary or wrong.

Also need to verify that `sample_weight_eval_set` still works correctly in 3.2.

## Scope

Files to investigate and update:
- `packages/openstef-models/src/openstef_models/utils/loss_functions.py` — objective functions (`arctan_loss_multi_objective`, `pinball_loss_multi_objective`, `xgb_prepare_target_for_objective`)
- `packages/openstef-beam/src/openstef_beam/metrics/metrics_probabilistic.py` — `mean_pinball_loss` (used as eval metric)
- `packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py` — `fit()` / `predict()` reshape logic

## Acceptance Criteria

- [ ] A minimal reproduction script that demonstrates the failure with XGBoost 3.2
- [ ] Objective functions handle both 1D (old) and 2D (new) input shapes, or exclusively use the 2D API
- [ ] All existing XGBoost forecaster tests pass with XGBoost 3.2 installed
- [ ] Version constraint in `packages/openstef-models/pyproject.toml` relaxed back to `>=3,<4` (or `>=3.2,<4` if minimum is bumped)
- [ ] `uv.lock` updated

## Notes

- XGBoost 3.2 changelog: https://xgboost.readthedocs.io/en/latest/changes/v3.2.0.html
- Advanced custom objectives guide: https://xgboost.readthedocs.io/en/latest/tutorials/advanced_custom_obj.html
- The `multi_strategy="one_output_per_tree"` param docs: https://xgboost.readthedocs.io/en/latest/tutorials/multioutput.html
- `_objective_decorator` in the sklearn interface wraps the custom fn and passes `dmatrix.get_label()` / weights to it — check what shape this returns in 3.2 for multi-output models

### PR DESCRIPTION
**Pull request by egordm**
## Summary

Adds regression tests verifying that `arctan_loss_multi_objective` and `pinball_loss_multi_objective` accept **2D `(n_samples, n_quantiles)` input arrays** as passed by XGBoost 3.2+, and produce results identical to the flat 1D arrays used in older versions.

## Context

XGBoost 3.2 changed the calling convention for custom objectives with multi-output trees: arrays are now passed as 2D `(n_samples, n_outputs)` instead of the flat 1D `(n_samples * n_outputs,)` used previously. This was tracked in #865 and fixed in #866.

## Relationship to #866

- These tests **intentionally FAIL** on the unfixed code (this branch's base, `release/v4.0.0` before #866).
- They will **pass** once #866 is merged.
- This PR should be merged **after** #866 to ensure CI stays green.

## Tests added

- `test_loss_fn__2d_input_matches_1d_input[pinball]` — pinball objective with 2D input
- `test_loss_fn__2d_input_matches_1d_input[arctan]` — arctan objective with 2D input
- `test_loss_fn__2d_input_with_sample_weights_matches_1d[pinball]` — pinball with 2D input + sample weights
- `test_loss_fn__2d_input_with_sample_weights_matches_1d[arctan]` — arctan with 2D input + sample weights

Closes #865 (when merged after #866)

### CODE REVIEW

**Review by bartpleiter — DISMISSED** [review]
Looks good to me, but please look at the sonar output

**Review by bartpleiter — DISMISSED** [review]
LGTM
